Source/DafnyLanguageServer/Workspace/DiagnosticPublisher.cs

```diff
@@ -29,6 +35,11 @@ public class DiagnosticPublisher : IDiagnosticPublisher {
         Version = document.Version,
         Diagnostics = document.Diagnostics.ToArray(),
       };
+      if (previouslyPublishedDiagnostics.TryGetValue(document.Uri, out var previousParams) && previousParams.Diagnostics.Equals(diagnosticParameters.Diagnostics)) {
+        return;
+      }
+
+      previouslyPublishedDiagnostics.AddOrUpdate(document.Uri, _ => diagnosticParameters, (_, _) => diagnosticParameters);
```

Comment:
Could we share this code with the code below?

Reply:
I don't want to. LSP diagnostics and Ghostness hints conceptually have little in common. The ghostness hints happen to reuse LSP's `Diagnostic` type but they only partially fill it, not sure why they don't send just a `Range` instead.

Reply:
I meant just the part that looks up in the dictionary and sends a message if there's no match; would it be bad to share that?

Reply:
I'll have a look in the next PR (that also changes this file and publishes another cached thing)